src/pal/src/map/map.cpp

```diff
@@ -2193,96 +2272,170 @@ MAPmmapAndRecord(
 
 Parameters:
     IN hFile - file to map
+    IN path - path to mapped file
+    OUT size - mapped virtual size
+    IN prevLoadedBase - previously loaded base
 
 Return value:
     non-NULL - the base address of the mapped image
     NULL - error, with last error set.
 --*/
 
-void * MAPMapPEFile(HANDLE hFile)
+void * MAPMapPEFile(HANDLE hFile, LPCWSTR path, size_t *size, void *prevLoadedBase)
 {
     PAL_ERROR palError = 0;
     IPalObject *pFileObject = NULL;
     IDataLock *pLocalDataLock = NULL;
     CFileProcessLocalData *pLocalData = NULL;
-    CPalThread *pThread = InternalGetCurrentThread();
-    void * loadedBase = NULL;
+    CPalThread *pThread;
+
+    bool doPreload = hFile == NULL;
+
     IMAGE_DOS_HEADER * loadedHeader = NULL;
     void * retval;
 #if _DEBUG
     bool forceRelocs = false;
     char* envVar;
 #endif
 
-    ENTRY("MAPMapPEFile (hFile=%p)\n", hFile);
+    size_t headerSize;
 
-    //Step 0: Verify values, find internal pal data structures.
-    if (INVALID_HANDLE_VALUE == hFile)
+    void * loadedBase = prevLoadedBase;
+    bool isPreloaded = loadedBase != NULL;
+    if (!((doPreload && prevLoadedBase == NULL && hFile == NULL && path != NULL && size != NULL)
+          || (!doPreload && hFile != NULL && path == NULL && size == NULL)))
```

Comment:
It would be great to split calculation of the expression into several steps with named intermediate values (like `isPreloaded`)